### PR TITLE
Fix perf regression in NuGet restore

### PR DESCRIPTION
**Pull request by ocallesp**
Avoid doing restore when receiving parts of the target frameworks. We can fix this by waiting for all target frameworks to get loaded and then publish.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/project-system/pull/8542)

### CODE REVIEW

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreUnconfiguredInputDataSource.cs [hunk 1]
```diff
@@ -40,13 +47,22 @@ protected override IDisposable LinkExternalInput(ITargetBlock<RestoreInfo> targe
...
                 _project);
 
-            // Transform all restore data -> combined restore data
-            DisposableValue<ISourceBlock<RestoreInfo>> mergeBlock = joinBlock.TransformWithNoDelta(update => update.Derive(MergeRestoreInputs));
+            IPropagatorBlock<IProjectVersionedValue<RestoreValues>, RestoreInfo> transformBlock
+                = DataflowBlockSlim.CreateTransformBlock<IProjectVersionedValue<RestoreValues>, RestoreInfo>(update => update.Derive(MergeRestoreInputs));
+
+            // Publish when all target frameworks get loaded
+            var mergeBlock = ProjectDataSources.SyncLinkTo(
+                _loadedActiveConfiguredProjectsDataSource.SourceBlock.SyncLinkOptions(),
```
Comment (ocallesp) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreUnconfiguredInputDataSource.cs:55]: _loadedActiveConfiguredProjectDataSource.SourceBlock is null.

Any idea on what can be wrong with this implementation ?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/ActiveConfiguredProjectsLoader.cs [hunk 2]
```diff
@@ -34,18 +33,16 @@ public ActiveConfiguredProjectsLoader(UnconfiguredProject project, IActiveConfig
...
-        {
             _subscription = _activeConfigurationGroupService.ActiveConfigurationGroupSource.SourceBlock.LinkTo(
                 target: _targetBlock,
                 linkOptions: DataflowOption.PropagateCompletion);
+
+            return Task.CompletedTask;
         }
 
+        public ITargetBlock<IProjectVersionedValue<IConfigurationGroup<ProjectConfiguration>>> TargetBlock => _targetBlock;
+
```
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/ActiveConfiguredProjectsLoader.cs:45]: Why this change?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/IActiveConfiguredProjectsLoadedDataSource.cs [hunk 3]
```diff
@@ -0,0 +1,8 @@
+﻿// Licensed to the .NET Foundation under one or more agreements. The .NET Foundation licenses this file to you under the MIT license. See the LICENSE.md file in the project root for more information.
+
+namespace Microsoft.VisualStudio.ProjectSystem;
+
+[ProjectSystemContract(ProjectSystemContractScope.UnconfiguredProject, ProjectSystemContractProvider.Private, Cardinality = Composition.ImportCardinality.ExactlyOne)]
+internal interface IActiveConfiguredProjectsLoadedDataSource : IProjectValueDataSource<IEnumerable<ConfiguredProject>>
+{
+}
```
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/IActiveConfiguredProjectsLoadedDataSource.cs:8]: Please add a summary for this description that explains what it publishes and why you might want to use it.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreUnconfiguredInputDataSource.cs [hunk 4]
```diff
@@ -40,13 +47,22 @@ protected override IDisposable LinkExternalInput(ITargetBlock<RestoreInfo> targe
...
+                joinBlock.SyncLinkOptions(),
+                target: transformBlock, 
+                linkOptions: DataflowOption.PropagateCompletion,
+                cancellationToken: CancellationToken.None
+                );
 
             JoinUpstreamDataSources(_activeConfigurationGroupService.ActiveConfiguredProjectGroupSource);
```
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreUnconfiguredInputDataSource.cs:62]: Do you also have to join `_loadedActiveConfiguredProjectsDataSource` here?

```suggestion
            JoinUpstreamDataSources(_activeConfigurationGroupService.ActiveConfiguredProjectGroupSource, _loadedActiveConfiguredProjectsDataSource);
```

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreUnconfiguredInputDataSource.cs [hunk 5]
```diff
@@ -40,13 +47,22 @@ protected override IDisposable LinkExternalInput(ITargetBlock<RestoreInfo> targe
...
 
             // Set the link up so that we publish changes to target block
-            mergeBlock.Value.LinkTo(targetBlock, DataflowOption.PropagateCompletion);
+            transformBlock.LinkTo(targetBlock, DataflowOption.PropagateCompletion);
 
             return new DisposableBag
```
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreUnconfiguredInputDataSource.cs:67]: Do you have to add `mergeBlock` to this disposable bag?

**Comment by ocallesp** [comment]
@drewnoakes I will implement this in a new class.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/IActiveConfiguredProjectsLoadedDataSource.cs (hunk 3 above)
Comment (ocallesp) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/IActiveConfiguredProjectsLoadedDataSource.cs:8]: I'll this in https://github.com/dotnet/project-system/pull/8543 as a cleanup commit

**Comment by ocallesp** [comment]
I'll add the corresponding fix to comments in this in https://github.com/dotnet/project-system/pull/8543 as a cleanup commit
